metaflow/plugins/aip/aip.py

```diff
@@ -294,6 +294,59 @@ def _create_workflow_yaml(
 
         KubeflowPipelines._add_archive_section_to_cards_artifacts(workflow)
 
+        if "onExit" in workflow["spec"]:
+            # replace entrypoint content with the exit handler handler content
+            """
+            # What it looks like beforehand...
+            entrypoint: helloflow
+            templates:
+            - name: exit-handler-1
+              dag:
+                tasks:
+                - name: end
+                  template: end
+                  dependencies: [start]
+                - {name: start, template: start}
+            - name: helloflow
+              dag:
+                tasks:
+                - {name: exit-handler-1, template: exit-handler-1}
+                - {name: sqs-exit-handler, template: sqs-exit-handler}
+            """
+            # find the exit-handler-1 template
+            exit_handler_template: dict = [
+                template
+                for template in workflow["spec"]["templates"]
+                if template["name"] == "exit-handler-1"
+            ][0]
+
+            # find the entrypoint template
+            entrypoint_template: dict = [
+                template
+                for template in workflow["spec"]["templates"]
+                if template["name"] == workflow["spec"]["entrypoint"]
+            ][0]
+
+            # replace the entrypoint template with the exit handler template
+            entrypoint_template["dag"] = exit_handler_template["dag"]
+
+            # rename exit-handler-1 to exit-handler
+            exit_handler_template["name"] = "exit-handler"
+            workflow["spec"]["onExit"] = "exit-handler"
+            exit_handler_template["dag"] = {
+                "tasks": [
+                    {
+                        "name": "sqs-exit-handler",
+                        "template": "sqs-exit-handler",
+                        "dependencies": ["notify-email-exit-handler"],
+                    },
+                    {
+                        "name": "notify-email-exit-handler",
+                        "template": "notify-email-exit-handler",
+                    },
+                ]
```

Comment:
Can we add tasks depending on configurations? If there's no sqs configured we might not need the sqs-exit-handler step. The same goes for email or other exit handlers we add in the future.

Reply:
The configuration is determined at runtime, and an be configured at runtime.

Reply:
I take that back.

* Whether to call the notify and SQS exit handlers is a compile time parameter.  I'll only add that step if the compile time parameter is included.
* Furthermore, I'll also use Argo `when` to only invoke the exit handler on workflow failure if `--notify-on-error` or always if `--notify-on-success`.  The same with the SQS exit handler, to only call on workflow failure and if `--sqs-url-on-error` is specified.